java/java.source.base/src/org/netbeans/modules/java/source/save/CasualDiff.java

```diff
@@ -1876,27 +1876,75 @@ protected int diffSwitch(JCSwitch oldT, JCSwitch newT, int[] bounds) {
 
     protected int diffCase(JCCase oldT, JCCase newT, int[] bounds) {
         int localPointer = bounds[0];
-        if (oldT.pat != null) {
-            int[] patBounds = getBounds(oldT.pat);
-            copyTo(localPointer, patBounds[0]);
-            localPointer = diffTree(oldT.pat, newT.pat, patBounds);
-            tokenSequence.move(patBounds[1]);
-            do { } while (tokenSequence.moveNext() && JavaTokenId.COLON != tokenSequence.token().id());
+        List<JCExpression> oldPatterns = getCasePatterns(oldT);
+        List<JCExpression> newPatterns = getCasePatterns(newT);
+        PositionEstimator patternEst = EstimatorFactory.casePatterns(
+                oldPatterns,
+                newPatterns,
+                diffContext
+        );
+        int posHint;
+        int endpos;
+        int copyTo;
+        if (oldPatterns.isEmpty()) {
+            moveFwdToOneOfTokens(tokenSequence, bounds[0], EnumSet.of(JavaTokenId.DEFAULT));
             tokenSequence.moveNext();
-            copyTo(localPointer, localPointer = tokenSequence.offset());
+            copyTo = endpos = posHint = tokenSequence.offset();
+
+            if (!newPatterns.isEmpty()) {
+                copyTo = getOldPos(oldT);
+            }
+        } else {
+            copyTo = posHint = oldPatterns.iterator().next().getStartPosition();
+            endpos = endPos(oldPatterns.get(oldPatterns.size() - 1));
+
+            if (newPatterns.isEmpty()) {
+                localPointer = copyTo = posHint = endpos;
+                printer.print("default");
+            }
         }
-        // todo (#pf): hot-fix of #113313, think about correct matching later
-        if (oldT.pat == null && newT.pat != null) {
-            printer.print(newT);
-            printer.newline();
-            return bounds[1];
+        copyTo(localPointer, copyTo);
+        localPointer = diffList2(oldPatterns, newPatterns, posHint, patternEst);
+        tokenSequence.move(endpos);
+        do { } while (tokenSequence.moveNext() && JavaTokenId.COLON != tokenSequence.token().id() && JavaTokenId.ARROW != tokenSequence.token().id());
```

Comment:
Maybe add some guard against incomplete / malformed source that is missing the case delimite so the loop does not pass too far.